src/query/storage/fanout/storage.go

```diff
@@ -91,35 +94,68 @@ func (s *fanoutStorage) FetchBlocks(
 		return stores[0].FetchBlocks(ctx, query, options)
 	}
 
-	// TODO(arnikola): use a genny map here instead, also execute in parallel.
-	blockResult := make(map[string]block.Block, 10)
+	var (
+		mu       sync.Mutex
+		wg       sync.WaitGroup
+		multiErr xerrors.MultiError
+	)
+
+	// TODO(arnikola): update this to use a genny map
+	blockResult := make(map[string]block.Block, len(stores))
+	wg.Add(len(stores))
 	for _, store := range stores {
-		result, err := store.FetchBlocks(ctx, query, options)
-		if err != nil {
-			return block.Result{}, err
-		}
+		store := store
+		go func() {
+			defer wg.Done()
+			result, err := store.FetchBlocks(ctx, query, options)
+			mu.Lock()
+			defer mu.Unlock()
 
-		for _, bl := range result.Blocks {
-			it, err := bl.SeriesIter()
 			if err != nil {
-				return block.Result{}, err
+				multiErr = multiErr.Add(err)
+				return
 			}
 
-			key := it.Meta().Bounds.String()
-			if foundBlock, found := blockResult[key]; found {
-				// this block exists. Check to see if it's already an appendable block.
-				if acc, ok := foundBlock.(block.AccumulatorBlock); ok {
-					// already an accumulator block, add current block.
-					if err := acc.AddBlock(bl); err != nil {
-						return block.Result{}, err
+			for _, bl := range result.Blocks {
+				key := bl.Meta().String()
+				foundBlock, found := blockResult[key]
+				if !found {
+					blockResult[key] = bl
+					continue
+				}
+
+				// This block exists. Check to see if it's already an appendable block.
+				// FIXME: (arnikola) use Block.Info() here to determine if it's an
+				// accumulator once #1888 merges.
+				blockType := foundBlock.Info().Type()
+				if blockType != block.BlockContainer {
+					var err error
+					blockResult[key], err = block.NewContainerBlock(foundBlock, bl)
+					if err != nil {
+						multiErr = multiErr.Add(err)
+						return
 					}
-				} else {
-					blockResult[key] = block.NewContainerBlock(foundBlock, bl)
+
+					continue
+				}
+
+				accumulator, ok := foundBlock.(block.AccumulatorBlock)
+				if !ok {
+					multiErr = multiErr.Add(fmt.Errorf("container block has incorrect type"))
```

Comment:
Hm, shouldn't this be a `return` after adding the error to it?

Reply:
Oops, nice catch